### PR TITLE
Discovering in shared wallets

### PR DESCRIPTION
**Pull request by paweljakubas**
# Issue Number

adp-927

# Overview

- [x] Adding the integration test showing the case
- [x] implement IsOurs and enable discovery
- [x] introduce hardened index guard outlawing improper ix in account when creating wallet
- [x] investigated another chain follower error for active shared wallet. It turn out there is very subtle error in sqlite handler when inserting the cosigner. Before that all cosigners were deleted which is incorrect, only cosigner row for a given credential and walid should be deleted. Otherwise, payment cosigners are deleted when delegation template is present and reading (selecting) is erroneous.

# Comments

### CODE REVIEW

**Comment by paweljakubas** [comment]
@piotr-iohk could you confirm the two issues in the ticket are fixed now, please :-)

File: lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Shared.hs [hunk 1]
```diff
@@ -481,8 +486,23 @@ isShared addr st = case fields st of
...
+    ) => IsOurs (SharedState n k) Address
+  where
+    isOurs addr state@(SharedState prefix _) =
+        let DerivationPrefix (purpose, coinType, accIx) = prefix
+            decorateIx addrIx = NE.fromList
+                [ DerivationIndex $ getIndex purpose
+                , DerivationIndex $ getIndex coinType
+                , DerivationIndex $ getIndex accIx
+                , DerivationIndex $ getIndex utxoExternal
+                , DerivationIndex $ getIndex addrIx
```
Comment (rvl) [inline lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Shared.hs:503]: This Index / DerivationIndex stuff ... not confusing at all!

File: lib/core/src/Cardano/Wallet/Api/Server.hs [hunk 2]
```diff
@@ -911,12 +913,12 @@ postSharedWalletFromRootXPrv ctx generateKey body = do
...
     rootXPrv = generateKey (seed, secondFactor) pwd
     g = defaultAddressPoolGap
-    accIx = Index $ getDerivationIndex $ getApiT (body ^. #accountIndex)
+    ix = getApiT (body ^. #accountIndex)
     wid = WalletId $ digest $ publicKey rootXPrv
     pTemplate = scriptTemplateFromSelf (getRawKey accXPub) $ body ^. #paymentScriptTemplate
     dTemplateM = scriptTemplateFromSelf (getRawKey accXPub) <$> body ^. #delegationScriptTemplate
     wName = getApiT (body ^. #name)
-    accXPub = publicKey $ deriveAccountPrivateKey pwd rootXPrv accIx
+    accXPub = publicKey $ deriveAccountPrivateKey pwd rootXPrv (Index $ getDerivationIndex ix)
```
Comment (rvl) [inline lib/core/src/Cardano/Wallet/Api/Server.hs:934]: Why this change?

**Review by rvl — APPROVED** [review]
Looks good.

File: lib/core/src/Cardano/Wallet/Api/Server.hs (hunk 2 above)
Comment (paweljakubas) [reply lib/core/src/Cardano/Wallet/Api/Server.hs:934]: this is just code reshuffling allowing to use guardHardened above. No logic was changed

**Comment by piotr-iohk** [comment]
> @piotr-iohk could you confirm the two issues in the ticket are fixed now, please :-)

Looks good now! :+1: Both cannot be reproduced.

**Comment by paweljakubas** [comment]
spotted in integration testing:
```

cardano-wallet                 &gt; [cardano-wallet.network:Warning:2092] [2021-05-28 18:40:41.48 UTC] Connection lost with the node. Network.Socket.recvBuf: resource vanished (Connection reset by peer)
--



cardano-wallet                 &gt;
--



cardano-wallet                 &gt; Failures:
--



cardano-wallet                 &gt;
--



cardano-wallet                 &gt;   src/Test/Integration/Scenario/API/Byron/Wallets.hs:187:59:
--



cardano-wallet                 &gt;   1) API Specifications, BYRON_WALLETS, BYRON_RESTORE_01, GET_01, LIST_01 - Restore a wallet, ledger 21 words
--



cardano-wallet                 &gt;        While verifying (Status {statusCode = 200, statusMessage = "OK"},Right (ApiByronWallet {id = ApiT {getApiT = WalletId {getWalletId = 7784a684aefeacf73e52ddd5746a1d0b7402347b}}, balance = ApiByronWalletBalance {available = Quantity {getQuantity = 0}, total = Quantity {getQuantity = 0}}, assets = ApiWalletAssetsBalance {available = ApiT {getApiT = TokenMap (fromList [])}, total = ApiT {getApiT = TokenMap (fromList [])}}, discovery = DiscoverySequential, name = ApiT {getApiT = WalletName {getWalletName = "Empty Byron Wallet"}}, passphrase = Just (ApiWalletPassphraseInfo {lastUpdatedAt = 2021-05-28 18:18:35.624318912 UTC}), state = ApiT {getApiT = Syncing (Quantity {getQuantity = Percentage {getPercentage = 1037 % 1250}})}, tip = ApiBlockReference {absoluteSlotNumber = ApiT {getApiT = SlotNo 2023}, slotId = ApiSlotId {epochNumber = ApiT {getApiT = EpochNo {unEpochNo = 40}}, slotNumber = ApiT {getApiT = SlotInEpoch {unSlotInEpoch = 23}}}, time = 2021-05-28 18:17:13.6 UTC, block = ApiBlockInfo {height = Quantity {getQuantity = 1000}}}}))
--


cardano-wallet                 &gt;        Waited longer than 90s to resolve action: "wallet is available and ready".
```

**Comment by paweljakubas** [comment]
bors retry

**Comment by ghost** [comment]
So I gotta be honest I don't know what a kit of this means. But long story short I was stoked bro get into the new atomic wall, stocking and interest. I just started transferred every thing alllbabsuddenbi get this error that won't me me konfbin. I tried and tried to get in but not even my key phrase was working and I took a f****** picture of it. I real installed it and the code worked so I thought I was gravy. Took be a couple transaction nrhat this was a different wall. Wtf is that' I was was straight up moving me tire wallets to atomic 4k worth. Next thing I know it all over the place. I  was pissed that's my day saving. I have no I'm still learning the black chae blockchain  I ant for shit doe knowledge on itdown yet. WTF HAPPEND. HOW DO I LABlE MY SEHIT SO THIS CANT HAPOEN AGAIN?

**Comment by ghost** [comment]
This scared me half to death.

**Comment by ghost** [comment]
I sent 9 house trying to figure this shit our blast night
